### DISCUSSION TITLE
Trouble with Database Trigger Accessing raw_app_meta_data

### DISCUSSION THREAD

**Discussion by martijnvanderpol** [supabase/supabase#17126] [Questions]
Hey Supabase community!

We're working on a project where we've defined multiple roles (user, supervisor, and admin), each with associated profile tables (UserProfile, SupervisorProfile, and AdminProfile) that relate to the main User object.

To automate profile creation, we've set up a database trigger that runs AFTER INSERT ON auth.users. We assign the role_id to **auth.users.raw_app_meta_data**. As I understand, raw_app_meta_data is better suited for this as it's intended for the application layer and can't be modified by the user.

However, we've run into an issue. When we try to read the role_id from auth.users.raw_app_meta_data and insert it into the users table, it doesn't work. But when we use raw_user_meta_data (which is different from raw_app_meta_data), it works as expected.

Could this be a policy issue where the trigger lacks permission to read from raw_app_meta_data?

For clarity, here's the function and trigger we've set up. Note that this is a test version where we're attempting to pull first_name and last_name just to see if the mechanism works:

```
CREATE OR REPLACE FUNCTION public.handle_new_user()
RETURNS TRIGGER
LANGUAGE plpgsql
SECURITY DEFINER SET search_path = public
AS $$
BEGIN
  INSERT INTO public.users (id, first_name, last_name)
  VALUES (NEW.id, NEW.raw_app_meta_data->>'first_name', NEW.raw_app_meta_data->>'last_name');
  RETURN NEW;
END;
$$;

CREATE TRIGGER on_auth_user_created
	AFTER INSERT ON auth.users
	FOR EACH ROW EXECUTE FUNCTION public.handle_new_user();
```

Any insights or suggestions are greatly appreciated! 🙏🏻

**Comment by GaryAustin1** [comment]
I believe the raw_app_meta_data column is now set with a 2nd update call.  So you would need to handle with an update trigger.

You can log out new with `raise log 'new = %',new;` and see the result in the Postgres log.
